src/core/thread/mle_router.cpp

```diff
@@ -3292,6 +3297,17 @@ void MleRouter::SendDataResponse(const Ip6::Address &aDestination,
         case Tlv::kPendingDataset:
             SuccessOrExit(error = AppendPendingDataset(*message));
             break;
+
+#if OPENTHREAD_CONFIG_LINK_METRICS_ENABLE
+        case Tlv::kLinkMetricsReport:
+            OT_ASSERT(aMessage != nullptr);
+            LinkMetricsQueryTlv linkMetricsQueryTlv;
+            SuccessOrExit(error = ThreadTlv::FindTlv(*aMessage, Tlv::kLinkMetricsQuery, sizeof(linkMetricsQueryTlv),
+                                                     linkMetricsQueryTlv));
+            SuccessOrExit(error =
+                              Get<LinkMetrics>().AppendLinkMetricsReport(*aMessage, &linkMetricsQueryTlv, *message));
```

Comment:
It looks strange to operate on another `Message` other than the sending message in this function. 
Why do we need so?

Reply:
New parameter`aMessage`(the message for received `MLE Data Request`) is introduced because in the `MLE Data Response`, the information of the incoming message is required for link metrics content.

`AppendLInkMetricsReport` handles the query content and appends report tlvs to the message. This is as a sub-step for preparing `MLE Data Response`.